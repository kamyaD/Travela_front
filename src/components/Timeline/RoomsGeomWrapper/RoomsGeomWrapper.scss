.rooms-geometry-wrapper {
  position: absolute;
  display: flex;
  flex-flow: column;
  width: 100%;
  height: 100%;
  padding-top: 36px;
  overflow: hidden;
  box-sizing: border-box;
}
