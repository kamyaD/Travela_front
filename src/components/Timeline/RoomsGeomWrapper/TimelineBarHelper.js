import moment from 'moment';
import getAveDaysPerMonth from '../helpers';

class TripGeomHelper {
  constructor({timelineStartDate, tripDayWidth, timelineViewType}) {
    this.values = {
      timelineStartDate,
      tripDayWidth,
      timelineViewType
    };
  }

  getStayPercentage = ({departureDate, returnDate, length, today}) => {
    // calcuate stay stayPercentage at minute accuracy
    let value, lengthInMinutes, minutesInDay=1440;
    switch (true) {
    case(today < departureDate):
      return '0%';
    case(today < returnDate):
      lengthInMinutes = length * minutesInDay;
      value = ((today.diff(departureDate, 'minutes') / lengthInMinutes) * 100);
      return `${value}%`;
    default:
      return '100%';
    }
  }

  calculateLeftOffsetError = (date) => {
    const {timelineViewType, timelineStartDate} = this.values;
    if(!(timelineViewType === 'year')) return 0;
    const monthsSince = date.diff(timelineStartDate, 'months');
    const daysInSegments = monthsSince * getAveDaysPerMonth(timelineStartDate);
    const actualDays = date.startOf('month').diff(
      timelineStartDate, 'days'
    );
    return (daysInSegments - actualDays);
  }

  getTripOffsetLeft = (departureDate) => {
    const { timelineStartDate, tripDayWidth } = this.values;
    const daysFromTimelineStart = departureDate.diff(timelineStartDate, 'days');
    const leftOffsetMarginErr = this.calculateLeftOffsetError(departureDate);
    return (daysFromTimelineStart * tripDayWidth) + leftOffsetMarginErr;
  }

  getTripStats = (trip) => {
    const today = moment();
    const departureDate = moment(trip.departureDate, 'YYYY-MM-DD');
    const returnDate = moment(trip.returnDate, 'YYYY-MM-DD');
    // last day of the trip inclusive
    const length = returnDate.add(1, 'days').diff(departureDate, 'days');
    const tripParams = {departureDate, returnDate, length, today};
    const stayPercentage = this.getStayPercentage(tripParams);
    const tripTimelineOffsetLeft = this.getTripOffsetLeft(departureDate);
    return { length, stayPercentage, tripTimelineOffsetLeft };
  }
}

export default TripGeomHelper;
