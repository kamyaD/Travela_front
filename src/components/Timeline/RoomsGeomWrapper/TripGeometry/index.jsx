import React, {Component, Fragment} from 'react';
import {PropTypes} from 'prop-types';
import moment from 'moment';
import TripDetails from './TripDetails';
import TripGeomHelper from './TripGeomHelper';
import {COLOR_SPEC_RANGES, SHADE_OFFSETS} from '../../settings';
import './TripGeometry.scss';

class TripGeometry extends Component {
  constructor(props) {
    super(props);
    this.state = {
      bookingDetailsVisible: false,
      colorSpecs: this.shuffleColorSpecs(),
      bookingDetailsPos: 0,
      translateDetailsLeft: false
    };
  }

  checkShouldTranslateDetails = (mouseX) => {
    const screenWidth = window.innerWidth;
    if(screenWidth/2 < mouseX)
      return true;
    return false;
  }

  setBookingDetailsAbsolutePos = (e) => {
    const tripGeomLeft = e.target.getBoundingClientRect().left;
    const mouseX = e.clientX;
    const bookingDetailsPos = mouseX - tripGeomLeft;
    const translateDetailsLeft = this.checkShouldTranslateDetails(mouseX);
    this.setState(prevState => ({
      ...prevState,
      bookingDetailsPos,
      translateDetailsLeft
    }));
  }

  toggleBookingDetails = (type) => {
    this.setState(prevState => {
      switch (type) {
      case 'open':
        if(prevState.bookingDetailsVisible)
          return prevState;
        return {
          ...prevState,
          bookingDetailsVisible: true
        };
      default:
        if(!prevState.bookingDetailsVisible)
          return prevState;
        return {
          ...prevState,
          bookingDetailsVisible: false
        };
      }
    });
  }

  getInnerColorShade = (colorSpecs) => {
    let {hue, saturation, lightness} = colorSpecs;
    saturation+=SHADE_OFFSETS.saturation;
    lightness+=SHADE_OFFSETS.lightness;
    return {hue, saturation, lightness};
  }

  pickValueFromRange = (colorSpec) => {
    // add a random fraction of the magnitude to min value
    return Math.round(colorSpec.min + (
      Math.random() * (colorSpec.max - colorSpec.min)
    ));
  }

  shuffleColorSpecs = () => {
    const hue = this.pickValueFromRange(COLOR_SPEC_RANGES.hue);
    const saturation = this.pickValueFromRange(COLOR_SPEC_RANGES.saturation);
    const lightness = this.pickValueFromRange(COLOR_SPEC_RANGES.lightness);
    return {hue, saturation, lightness};
  }

  getBookingDetailsVariant = () => {
    const {bookingDetailsVisible} = this.state;
    return bookingDetailsVisible ? 'visible' : 'hidden';
  }

  renderTripBarInnerElements(tripStats, colorSpecs) {
    const {trip} = this.props;
    const {hue, saturation, lightness} = this.getInnerColorShade(colorSpecs);
    const travellerName = trip.request.name;
    const travellerGenderLabel = trip.request.gender.charAt(0);
    return (
      <Fragment>
        <div
          className="geom-trip geom-trip--inner"
          style={{
            width: tripStats.stayPercentage,
            background: `hsl(${hue},${saturation}%,${lightness}%)`
          }}
        />
        <span>{`${travellerName}  (${travellerGenderLabel})`}</span>
      </Fragment>
    );
  }

  renderTripBar = (tripStats) => {
    const {colorSpecs} = this.state;
    const {tripDayWidth} = this.props;
    const {hue, saturation, lightness} = colorSpecs;
    return (
      <div
        className="geom-trip geom-trip--outer"
        style={{
          minHeight: '20px',
          background: `hsl(${hue},${saturation}%,${lightness}%)`,
          width: `${tripDayWidth * tripStats.length}px`
        }}
      >
        {this.renderTripBarInnerElements(tripStats, colorSpecs)}
      </div>
    );
  }

  render() {
    const {bookingDetailsPos, translateDetailsLeft} = this.state;
    const {trip, handleChangeRoomModal} = this.props;
    const detailsVariant = this.getBookingDetailsVariant();
    const helper = new TripGeomHelper(this.props);
    const tripStats = helper.getTripStats(trip);
    return (
      <div
        className="timeline-trip-geometry"
        tabIndex="-1"
        role="presentation"
        onKeyUp={()=>{}}
        onClick={this.setBookingDetailsAbsolutePos}
        onFocus={() => this.toggleBookingDetails('open')}
        onBlur={() => this.toggleBookingDetails('close')}
        style={{
          left: `${tripStats.tripTimelineOffsetLeft}px`
        }}
      >
        {this.renderTripBar(tripStats)}
        <TripDetails
          trip={trip}
          bookingDetailsPos={bookingDetailsPos}
          translateDetailsLeft={translateDetailsLeft}
          detailsVariantClass={detailsVariant}
          toggleBookingDetails={this.toggleBookingDetails}
          handleChangeRoomModal={handleChangeRoomModal}
        />
      </div>
    );
  }
}

TripGeometry.propTypes = {
  trip: PropTypes.object.isRequired,
  tripDayWidth: PropTypes.number.isRequired,
  handleChangeRoomModal: PropTypes.func.isRequired
};

export default TripGeometry;
