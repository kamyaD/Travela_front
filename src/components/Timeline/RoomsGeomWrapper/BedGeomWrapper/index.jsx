import React, {Component} from 'react';
import {PropTypes} from 'prop-types';
import moment from 'moment';
import TripGeometry from '../TripGeometry';

class BedGeomWrapper extends Component {
  render() {
    const {tripDayWidth, trips, timelineStartDate,
      timelineViewType, handleChangeRoomModal} = this.props;
    const tripGeoms = trips.map(trip => (
      <TripGeometry
        key={trip.id}
        tripDayWidth={tripDayWidth}
        timelineStartDate={timelineStartDate}
        trip={trip}
        timelineViewType={timelineViewType}
        handleChangeRoomModal={handleChangeRoomModal}
      />
    ));
    return (
      <div className="bed-geometry-wrapper item-row">
        {tripGeoms}
      </div>
    );
  }
}

BedGeomWrapper.propTypes = {
  tripDayWidth: PropTypes.number.isRequired,
  trips: PropTypes.array.isRequired,
  timelineStartDate: PropTypes.object.isRequired,
  timelineViewType: PropTypes.string,
  handleChangeRoomModal: PropTypes.func.isRequired
};

BedGeomWrapper.defaultProps = {
  timelineViewType: 'month'
};

export default BedGeomWrapper;
