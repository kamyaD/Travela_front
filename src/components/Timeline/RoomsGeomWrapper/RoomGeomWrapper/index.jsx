import React, {PureComponent} from 'react';
import {PropTypes} from 'prop-types';
import BedGeomWrapper from '../BedGeomWrapper';
import MaintenanceDetails from './MaintenanceDetails';
import TimelineBarWrapper from '../TimelineBarWrapper';
import TimelineBar from '../TimelineBarWrapper/TimelineBar';
import TimelineBarHelper from '../TimelineBarHelper';
import './RoomGeomWrapper.scss';

class RoomGeomWrapper extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      bookingDetailsVisible: false,
      bookingDetailsPos: 0,
      translateDetailsLeft: false
    };
  }

  getBookingDetailsVariant = () => {
    const {bookingDetailsVisible} = this.state;
    return bookingDetailsVisible ? 'visible' : 'hidden';
  }

  checkShouldTranslateDetails = (mouseX) => {
    const screenWidth = window.innerWidth;
    if(screenWidth/2 < mouseX)
      return true;
    return false;
  }

  setBookingDetailsAbsolutePos = (e) => {
    const tripGeomLeft = e.target.getBoundingClientRect().left;
    const mouseX = e.clientX;
    const bookingDetailsPos = mouseX - tripGeomLeft;
    const translateDetailsLeft = this.checkShouldTranslateDetails(mouseX);
    this.setState(prevState => ({
      ...prevState,
      bookingDetailsPos,
      translateDetailsLeft
    }));
  }

  toggleBookingDetails = (type) => {
    this.setState(prevState => {
      switch (type) {
      case 'open':
        if(prevState.bookingDetailsVisible)
          return prevState;
        return {
          ...prevState,
          bookingDetailsVisible: true
        };
      default:
        if(!prevState.bookingDetailsVisible)
          return prevState;
        return {
          ...prevState,
          bookingDetailsVisible: false
        };
      }
    });
  }

  renderMaintananceBar = (tripStats) => {
    const {tripDayWidth} = this.props;
    return (
      <TimelineBar
        tripDayWidth={tripDayWidth}
        tripStats={tripStats}
        customStyle={{
          background: '#FF5359',
        }}
      >
        <div className="maintainance-bar inner">
          Unavailable
        </div>
      </TimelineBar>
    );
  }

  renderMaintenanceDetails = maintenance => {
    const {bookingDetailsPos, translateDetailsLeft} = this.state;
    const detailsVariant = this.getBookingDetailsVariant();
    const {handleDeleteMaintenanceModal, handleEditMaintenanceModal} = this.props;
    return (
      <MaintenanceDetails
        bookingDetailsPos={bookingDetailsPos}
        translateDetailsLeft={translateDetailsLeft}
        toggleBookingDetails={this.toggleBookingDetails}
        detailsVariantClass={detailsVariant}
        handleEditMaintenanceModal={handleEditMaintenanceModal}
        handleDeleteMaintenanceModal={handleDeleteMaintenanceModal}
        maintenance={maintenance}
      />
    );
  }
  
  render() {
    const {beds, timelineStartDate, tripDayWidth,timelineViewType,
      handleChangeRoomModal, maintainances, deletedMaintenance} = this.props;
    const currentMaintenances = (deletedMaintenance.roomId) 
      ? maintainances.filter(room => room.roomId !== deletedMaintenance.roomId) 
      : maintainances;

    const bedGeoms = beds.map(bed => (
      <BedGeomWrapper
        key={bed.id}
        trips={bed.trips}
        timelineStartDate={timelineStartDate}
        tripDayWidth={tripDayWidth}
        timelineViewType={timelineViewType}
        handleChangeRoomModal={handleChangeRoomModal}
      />
    ));
    const maintainanceBars = currentMaintenances.map(maintainance => {
      maintainance.departureDate = maintainance.start;
      maintainance.returnDate = maintainance.end;
      const helper = new TimelineBarHelper(this.props);
      const tripStats = helper.getTripStats(maintainance);
      return (
        <TimelineBarWrapper
          toggleBookingDetails={this.toggleBookingDetails}
          setBookingDetailsAbsolutePos={this.setBookingDetailsAbsolutePos}
          tripStats={tripStats}
          key={maintainance.id}
        >
          {this.renderMaintananceBar(tripStats)}
          {this.renderMaintenanceDetails(maintainance)}
        </TimelineBarWrapper>
      );
    });
    return (
      <div className="room-geometry-wrapper">
        <div className="room-status-bar item-row">
          {maintainanceBars}
        </div>
        {bedGeoms}
      </div>
    );
  }
}

RoomGeomWrapper.propTypes = {
  deletedMaintenance: PropTypes.object,
  maintainances: PropTypes.array.isRequired,
  beds: PropTypes.array.isRequired,
  timelineStartDate: PropTypes.object.isRequired,
  tripDayWidth: PropTypes.number.isRequired,
  timelineViewType: PropTypes.string,
  handleChangeRoomModal: PropTypes.func.isRequired,
  handleEditMaintenanceModal: PropTypes.func.isRequired,
  handleDeleteMaintenanceModal: PropTypes.func.isRequired
};

RoomGeomWrapper.defaultProps = {
  deletedMaintenance: {},
  timelineViewType: 'month'
};

export default RoomGeomWrapper;
