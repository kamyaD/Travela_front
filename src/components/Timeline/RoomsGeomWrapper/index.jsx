import React, {PureComponent} from 'react';
import {PropTypes} from 'prop-types';
import RoomGeomWrapper from './RoomGeomWrapper';
import './RoomsGeomWrapper.scss';

class RoomsGeomWrapper extends PureComponent {
  render() {
    const {
      rooms, timelineStartDate, tripDayWidth, handleDeleteMaintenanceModal, deletedMaintenance,
      timelineViewType, handleChangeRoomModal, handleEditMaintenanceModal, editMaintenance
    } = this.props;
    const roomGeoms = rooms.map(room => (
      <RoomGeomWrapper
        status={room.faulty}
        key={room.id}
        beds={room.beds}
        editMaintenance={editMaintenance}
        deletedMaintenance={deletedMaintenance}
        maintainances={room.maintainances}
        timelineStartDate={timelineStartDate}
        tripDayWidth={tripDayWidth}
        timelineViewType={timelineViewType}
        handleChangeRoomModal={handleChangeRoomModal}
        handleEditMaintenanceModal={handleEditMaintenanceModal}
        handleDeleteMaintenanceModal={handleDeleteMaintenanceModal} />
    ));
    return (<div className="rooms-geometry-wrapper">{roomGeoms}</div>);
  }
}

RoomsGeomWrapper.propTypes = {
  editMaintenance: PropTypes.object,
  deletedMaintenance: PropTypes.object,
  rooms: PropTypes.array.isRequired,
  timelineStartDate: PropTypes.object.isRequired,
  tripDayWidth: PropTypes.number.isRequired,
  timelineViewType: PropTypes.string,
  handleChangeRoomModal: PropTypes.func.isRequired,
  handleEditMaintenanceModal: PropTypes.func.isRequired,
  handleDeleteMaintenanceModal: PropTypes.func.isRequired
};

RoomsGeomWrapper.defaultProps = {
  timelineViewType: 'month',
  editMaintenance: {},
  deletedMaintenance: {}
};

export default RoomsGeomWrapper;
