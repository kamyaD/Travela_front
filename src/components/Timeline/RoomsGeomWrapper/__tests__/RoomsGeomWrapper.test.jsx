import React from 'react';
import moment from 'moment';
import RoomsGeomWrapper from '..';

const props = {
  rooms: [{
    faulty: false,
    id: 'room-id-1',
    beds: [{
      bedId: 'kitanda1',
      bedName: 'kitanda 1'
    }],
    maintainances: [{
      reason: 'Bad window',
      start: '11-10-2018',
      end: '11-10-2018'
    }],
  }],
  timelineStartDate: moment().startOf('month'),
  tripDayWidth: 31,
  timelineViewType: 'month',
  handleChangeRoomModal: jest.fn(),
  handleEditMaintenanceModal: jest.fn(),
  handleDeleteMaintenanceModal: jest.fn()
};

describe('<RoomsGeomWrapper />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <RoomsGeomWrapper {...props} />
    );
  });

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a wrapper for all rooms\' visualization geometries', () => {
    expect(wrapper.find('.rooms-geometry-wrapper')).toHaveLength(1);
  });

  it('renders the RoomGeomWrapper for the rooms', () => {
    const roomGeomWrappers = wrapper.find('RoomGeomWrapper');
    expect(roomGeomWrappers).toHaveLength(1);
  });
});
