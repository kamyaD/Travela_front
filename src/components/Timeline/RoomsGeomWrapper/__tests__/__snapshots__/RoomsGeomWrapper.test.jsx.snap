// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomsGeomWrapper /> renders properly 1`] = `
<div
  className="rooms-geometry-wrapper"
>
  <RoomGeomWrapper
    beds={
      Array [
        Object {
          "bedId": "kitanda1",
          "bedName": "kitanda 1",
        },
      ]
    }
    deletedMaintenance={Object {}}
    editMaintenance={Object {}}
    key="room-id-1"
    status={false}
    timelineStartDate={"2018-12-01T00:00:00.000Z"}
    timelineViewType="month"
    tripDayWidth={31}
  />
</div>
`;
